### PR TITLE
Update 9-design.md

### PR DESCRIPTION
**Pull request by achillet3**

### CODE REVIEW

File: template/9-design.md [hunk 1]
```diff
@@ -1,56 +1,193 @@
 # Design and Implementation
 
 ## Frontend
+### Implementation Framework
+The app will be developed in Kotlin, optimized for Android, utilizing the Model-View-ViewModel (MVVM)architecture. This approach ensures a clear separation of concerns, making the management of the app’s UI and business logic more efficient and maintainable.
```
Comment (christinehahaa) [inline template/9-design.md]:
```suggestion
The app will be developed in Kotlin, optimized for Android, utilizing the Model-View-ViewModel (MVVM)^architecture. This approach ensures a clear separation of concerns, making the management of the app’s UI and business logic more efficient and maintainable.
```

File: template/9-design.md [hunk 2]
```diff
@@ -1,56 +1,193 @@
...
+### Key Libraries and Languages
+- **Kotlin**: Primary language for our Android development
+- **Hilt**: Dependency injection library for managing dependencies and improving code modularity.
+- **Material3**: Used for all graphic components
+- **Jetpack Compose**: Used for dynamic UI rendering.
+- **Firebase**: Backend-as-a-Service platform providing authentication, database, and cloud storage.
+- **Android Camera Image Analysis**: Utilized for integrating camera functionalities required for photo capture during friend requests.
+- **Dagger Assisted Factory**: Design pattern to create objects, particularly useful in the ViewModel creation process, ensuring that instances are correctly initialized with necessary dependencies.
+- **ZXing**: Library to encode information to QR codes and decode QR codes to information
```
Comment (christinehahaa) [inline template/9-design.md:14]: Not sure if we should add Nominatim API for location searching?

File: template/9-design.md [hunk 3]
```diff
@@ -1,56 +1,193 @@
...
+### Dynamic UI Rendering Strategy
+The app will leverage Jetpack Compose for dynamic UI rendering. Compose enables the creation of UI components declaratively, allowing for real-time updates and modifications based on data changes.
+- **Composable Functions**: These will be used to define each screen and UI component, enabling reusable and modular UI elements.
+- **State Management**: ViewModels will manage the state, ensuring the UI is always in sync with the underlying data model.
+### Essential Screens
+#### 1 Home Screen
+- **Features**: Displays all events as well as events the user has registered to in either list or map view
+- **Components**: Event list, Upcoming events, search-filter, list/map view button
```
Comment (christinehahaa) [inline template/9-design.md:28]: perhaps mention nav bar?

File: template/9-design.md [hunk 4]
```diff
@@ -1,56 +1,193 @@
...
-- Firestore Storage
-
-*How is it shared/copied/cached?*
-- Utilize Firestore’s built-in caching on the client side to reduce latency and enhance the user experience by serving data from cache.
+### Data Collection and Management
+#### User Data
+- **Names:** User's first and last names. (String)
+- **Email Addresses:** User's email for authentication and communication. (String)
+- **Friends:** List of friends connected within the app. (Array of userIds)
+- **Bio:** Short presentation text. (String)
```
Comment (nikhen-s) [inline template/9-design.md]:
```suggestion
- **Bio:** Short description of a user. (String)
```

File: template/9-design.md [hunk 5]
```diff
@@ -1,56 +1,193 @@
...
+- **Features**: Enables users to create new events with detailed specifications
+- **Components**: Input Fields for Event Details (Name, Photo, Description, Category, Start/End Date & Time, Location, Ticket Info), and a create event button
+- **Description**: The Event Creation Screen simplifies the event creation process, providing users with input fields for all necessary event details. The interface encourages users to organize and host events.
+#### 3 User Profile Screen
+- **Features**: Displays the user's personal information for viewing and editing.
+- **Components**: User's Personal Information Input Fields (Text fields or input fields, depending on edit mode) and edit button
+- **Description**: The User Profile Screen offers a view of the user's information, allowing easy access for review and modification. 
+#### 4 Messages Screen
+- **Features**: Facilitates real-time messaging between users.
+- **Components**: Chat list and My Friends’ chat list
```
Comment (christinehahaa) [inline template/9-design.md]:
```suggestion
- **Components**: Chat list and My Friends list
```

File: template/9-design.md [hunk 6]
```diff
@@ -1,56 +1,193 @@
...
-
-*Where is it stored?*
-- Firebase Database
-- Firestore Storage
-
-*How is it shared/copied/cached?*
-- Utilize Firestore’s built-in caching on the client side to reduce latency and enhance the user experience by serving data from cache.
+### Data Collection and Management
+#### User Data
+- **Names:** User's first and last names. (String)
```
Comment (nikhen-s) [inline template/9-design.md]: Nit: Maybe we could change all of these from plural to singular: e.g. Names to name, email addresses to email address, etc to ensure consistency.

File: template/9-design.md [hunk 7]
```diff
@@ -1,56 +1,193 @@
...
+  - **Location lat**: Latitude of the location (Number)
+  - **Location lng**: Longitude of the location (Number)
+- **Ticket Information**: Details of the ticket
+  - **Ticket name**: Name of the type of the ticket (String)
+  - **Ticket price**: Price of the type of the ticket (Number)
+  - **Ticket capacity**: Number of available tickets of the type of the ticket (Number)
+  - **Ticket purchases**: Number of reserved tickets of the type of the ticket (Number)
+- **Attendee List**: List of users participating in events. (Array of userIds)
+- **Main Organizer**: Id of the user organizing the event. (String)
+- **Organizer List**: List of users organizing events. (Array of userIds)
```
Comment (nikhen-s) [inline template/9-design.md]:
```suggestion
- **Organizer List**: List of user ids of users organizing events. (Array of userIds)
```

File: template/9-design.md [hunk 8]
```diff
@@ -1,56 +1,193 @@
...
+- **Description**: The Event Creation Screen simplifies the event creation process, providing users with input fields for all necessary event details. The interface encourages users to organize and host events.
+#### 3 User Profile Screen
+- **Features**: Displays the user's personal information for viewing and editing.
+- **Components**: User's Personal Information Input Fields (Text fields or input fields, depending on edit mode) and edit button
+- **Description**: The User Profile Screen offers a view of the user's information, allowing easy access for review and modification. 
+#### 4 Messages Screen
+- **Features**: Facilitates real-time messaging between users.
+- **Components**: Chat list and My Friends’ chat list
+- **Description**: The Messaging Screen serves as a central hub for users to engage in real-time conversations. It provides a comprehensive view of all conversations, allowing users to seamlessly switch between chats. Additionally, a tab allows users to specifically view conversations with their friends, streamlining communication with close contacts.
```
Comment (christinehahaa) [inline template/9-design.md]:
```suggestion
- **Description**: The Messaging Screen serves as a central hub for users to engage in real-time conversations. It provides a comprehensive view of all conversations, allowing users to seamlessly switch between chats. Additionally, a tab allows users to specifically view their friends list, streamlining the ability to review the connections made and communications with close contacts.
```

File: template/9-design.md [hunk 9]
```diff
@@ -1,56 +1,193 @@
...
+- Continuous Integration/Continuous Deployment (CI/CD) pipelines are set up to automate the testing and deployment processes, allowing for rapid iterations and updates.
+### Deployment
+- The application is deployed on cloud platforms like AWS or Azure to ensure high availability and scalability.
+- Special attention is paid to infrastructure requirements to support the expected load and user base of the application.
+- Firebase Authentication is utilized to secure APIs, ensuring that data requests are authenticated and users can only access data relevant to them.
+- HTTPS is enforced for all communications between the app and servers to guarantee the confidentiality and integrity of data transmission.
+### Special Infrastructure Requirements
+- Scalable backend infrastructure is provisioned to handle potential spikes in traffic, especially during peak usage periods or major events.
+- Load balancing mechanisms are implemented to distribute traffic evenly across multiple server instances, optimizing performance and reliability.
```
Comment (nikhen-s) [inline template/9-design.md]: Do we need load balancing for the MVP initially? Maybe this is when we have a larger user base.

File: template/9-design.md [hunk 10]
```diff
@@ -1,56 +1,193 @@
...
-- Home screen, Event Details, User Profile, Messaging, Event Creation.
 ## Backend
-
-*Decompose the MVP into functional blocks.*
-- User Management: Handles registrations, profiles, and settings.
-- Event Management: Creates and updates events, manages tickets.
-- Messaging System: Manages communication between users.
+### User Management
+- **Description**: This functional block handles user-related operations such as registrations, profiles, and interactions between users (e.g., becoming friends).
+- **Application Logic**: Manages user authentication, registration, profile updates, and friend requests.
```
Comment (christinehahaa) [inline template/9-design.md]:
```suggestion
- **Application Logic**: Manages user authentication, registration, profile updates, and adding friends via QR scan.
```

File: template/9-design.md [hunk 11]
```diff
@@ -1,56 +1,193 @@
...
-*Any special infrastructure requirements.*
-- Utilizing hilt for dependency injection in testing.
\ No newline at end of file
+### Testing Strategies
+#### UI Tests
+- UI tests are conducted to verify that UI components, including those instantiated by factories, run smoothly and interact correctly.
+- Tests cover UI elements such as buttons, input fields, dialogs, etc., instantiated by factories.
+#### ViewModel Tests
+- ViewModel tests ensure that appropriate states are prepared for UI components instantiated by factories.
+- Tests verify the logic and behavior of ViewModel classes, including interactions with factory-created UI components.
```
Comment (nikhen-s) [inline template/9-design.md:188]:
```suggestion
- Tests verify the logic and behavior of ViewModel classes, including interactions with factory-created UI components.
- Tests utilize Mock Repositories for the databases to isolate testing specifically on the ViewModel.
```

File: template/9-design.md (hunk 2 above)
Comment (christinehahaa) [inline template/9-design.md:14]: also need to add something about 3rd party payment api?

**Review by nikhen-s — CHANGES_REQUESTED** [review]
Just small nits, other than that LGTM, well done!

File: template/9-design.md [hunk 12]
```diff
@@ -1,56 +1,193 @@
...
+- **Description**: This functional block handles user-related operations such as registrations, profiles, and interactions between users (e.g., becoming friends).
+- **Application Logic**: Manages user authentication, registration, profile updates, and friend requests.
+- **Framework**: Developed using Firebase Authentication and Firestore for data storage.
+- **Dynamic View Rendering**: Integrates with the frontend to render user profiles and friend lists dynamically based on data from Firestore.
+- **Responsive Backend-Frontend Interaction**: Ensures responsive interaction with the frontend, updating user profiles and friend lists based on user interactions.
+### Event Management
+- **Description**: This block handles operations related to creating, updating, and managing events, including ticket management.
+- **Application Logic**: Manages event creation, updates, and ticket sales.
+- **Framework**: Utilizes Firebase Firestore for event data storage.
+- **Dynamic View Rendering**: Provides APIs to fetch event details dynamically, ensuring frontend displays reflect the latest event information.
```
Comment (christinehahaa) [inline template/9-design.md:80]: add something about payment API?

**Review by christinehahaa — CHANGES_REQUESTED** [review]
Some minor comments here and there, overall awesome work.

**Review by nikhen-s — APPROVED** [review]
Just approving so after changes, you can directly merge to main! Thank you!

File: template/9-design.md (hunk 6 above)
Comment (achillet3) [reply template/9-design.md]: You're right! Everything was corrected

File: template/9-design.md (hunk 9 above)
Comment (achillet3) [reply template/9-design.md]: I have taken it out as there shouldn't too much issues in the beginning indeed

File: template/9-design.md (hunk 3 above)
Comment (achillet3) [reply template/9-design.md:28]: Added it as an extra Essential Screen and explained that it was at the bottom of all other screens

File: template/9-design.md (hunk 2 above)
Comment (achillet3) [reply template/9-design.md:14]: You're right! Added a third-party payment SDK to integrate this

File: template/9-design.md (hunk 12 above)
Comment (achillet3) [reply template/9-design.md:80]: Modified as well to take that into account

**Review by wisdomseeker2001 — APPROVED** [review]

**Pull request closed without merging**